File: plugin_files/bin/scripting/hooks.lua
local msgpack_pack = msgpack.pack

function AddHook(mem, args_list, return_type)
    if type(args_list) ~= "string" then
        print("AddHook: args_list must be a string.")
        return
    end
    if type(return_type) ~= "string" then
        print("AddHook: return_type must be a string.")
        return
    end

    local hookId = hooks:Add(mem, args_list, return_type)

    if hookId == "00000000-0000-0000-0000-000000000000" then
        return
    end

    return { hookId = hookId }
end

function AddPreHookListener(hookHandle, callback)
    if type(hookHandle) ~= "table" then
        print("Error: Invalid hook data passed to AddPreHookListener.")
        return
    end

    if type(hookHandle.hookId) ~= "string" then
        print("Error: Invalid hook data passed to AddPreHookListener.")
        return
    end

    return AddEventHandler("hook:Pre:"..hookHandle.hookId, callback)
end

function AddPostHookListener(hookHandle, callback)
    if type(hookHandle) ~= "table" then
        print("Error: Invalid hook data passed to AddPostHookListener.")
        return
    end

    if type(hookHandle.hookId) ~= "string" then
        print("Error: Invalid hook data passed to AddPostHookListener.")
        return
    end

    return AddEventHandler("hook:Post:"..hookHandle.hookId, callback)
end

function RemoveHookListener(hookListener)
    if type(hookListener) ~= "table" then
        print("Error: Invalid hook data passed to RemoveHookListener.")
        return
    end

    if not hookListener.key or not hookListener.name then
        print("Error: Invalid hook data passed to RemoveHookListener.")
        return
    end
    
    RemoveEventHandler(hookListener)
end

function CallHook(hookHandle, ...)
    if type(hookHandle) ~= "table" then
        print("Error: Invalid hook data passed to AddPostHookListener.")
        return
    end

    if type(hookHandle.hookId) ~= "string" then
        print("Error: Invalid hook data passed to AddPostHookListener.")
        return
    end

    return hooks:Call(hookHandle.hookId, msgpack_pack({...}))
end

File: src/managed/API/Scripting/Hooks.cs
using Microsoft.Extensions.DependencyModel;
using SwiftlyS2.Internal_API;

namespace SwiftlyS2.API.Scripting
{
    public struct HookHandler
    {
        public string HookId;
    }

    public class Hooks
    {
        private static IntPtr _ctx = IntPtr.Zero;

        private static void InitializeContext()
        {
            if (_ctx != IntPtr.Zero) return;
            _ctx = Invoker.CallNative<IntPtr>("Hooks", "Hooks", CallKind.ClassFunction);
        }

        public static HookHandler AddHookEntityOutput(string class_name, string output)
        {
            InitializeContext();
            HookHandler hookHandler = new();
            hookHandler.HookId = Invoker.CallNative<string>("Hooks", "AddEntityOutputHook", Internal_API.CallKind.CoreClassFunction, _ctx, class_name, output) ?? "";
            return hookHandler;
        }

        public static HookHandler AddHook(Memory mem, string args_list, string return_type)
        {
            InitializeContext();
            HookHandler hookHandler = new();
            hookHandler.HookId = Invoker.CallNative<string>("Hooks", "Add", Internal_API.CallKind.CoreClassFunction, _ctx, mem, args_list, return_type) ?? "";
            return hookHandler;
        }

        public static HookHandler AddVHook(string library, string vtable_name, string offset, string args_list, string return_type)
        {
            InitializeContext();
            HookHandler hookHandler = new();
            hookHandler.HookId = Invoker.CallNative<string>("Hooks", "AddVirtual", Internal_API.CallKind.CoreClassFunction, _ctx, library, vtable_name, offset, args_list, return_type) ?? "";
            return hookHandler;
        }

        public static EventHandler AddPreHookListener(HookHandler handle, Func<Events.Event, EventResult?> cb)
        {
            return Events.AddEventHandler($"hook:Pre:{handle.HookId}", cb);
        }
        
        public static EventHandler AddPostHookListener(HookHandler handle, Func<Events.Event, EventResult?> cb)
        {
            return Events.AddEventHandler($"hook:Post:{handle.HookId}", cb);
        }

        public static void RemoveHookListener(EventHandler ev)
        {
            Events.RemoveEventHandler(ev);
        }

        public static void CallHook(HookHandler hookHandler, params object[] args)
        {
            InitializeContext();
            Invoker.CallNative("Hooks", "Call", Internal_API.CallKind.CoreClassFunction, _ctx, hookHandler.HookId, (object)args);
        }

        public static unsafe T? CallHook<T>(HookHandler hookHandler, params object[] args)
        {
            InitializeContext();
            IntPtr val = Internal_API.Invoker.CallNative<IntPtr>("Hooks", "Call", Internal_API.CallKind.CoreClassFunction, _ctx, hookHandler.HookId, (object)args);
            byte* p = (byte*)val;
            byte* tp = (byte*)&p;
            Type t = typeof(T);
            return (T?)CallContext.ReadValue(ref t, ref tp);
        }
    }
}